### PR TITLE
[SPARK-36477][SQL] Inferring schema from JSON file shall handle CharConversionException/MalformedInputException

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

When set `ignoreCorruptFiles=true`, reading JSON still fails with corrupt files during inferring schema.

```scala
java.io.CharConversionException: Unsupported UCS-4 endianness (2143) detected
	at com.fasterxml.jackson.core.json.ByteSourceJsonBootstrapper.reportWeirdUCS4(ByteSourceJsonBootstrapper.java:504)
	at com.fasterxml.jackson.core.json.ByteSourceJsonBootstrapper.checkUTF32(ByteSourceJsonBootstrapper.java:471)
	at com.fasterxml.jackson.core.json.ByteSourceJsonBootstrapper.detectEncoding(ByteSourceJsonBootstrapper.java:144)
	at com.fasterxml.jackson.core.json.ByteSourceJsonBootstrapper.constructParser(ByteSourceJsonBootstrapper.java:247)
	at com.fasterxml.jackson.core.JsonFactory._createParser(JsonFactory.java:1528)
	at com.fasterxml.jackson.core.JsonFactory.createParser(JsonFactory.java:1034)
	at org.apache.spark.sql.catalyst.json.CreateJacksonParser$.internalRow(CreateJacksonParser.scala:86)
	at org.apache.spark.sql.execution.datasources.json.TextInputJsonDataSource$.$anonfun$inferFromDataset$4(JsonDataSource.scala:107)
	at org.apache.spark.sql.catalyst.json.JsonInferSchema.$anonfun$infer$3(JsonInferSchema.scala:66)
	at org.apache.spark.util.Utils$.tryWithResource(Utils.scala:2621)
	at org.apache.spark.sql.catalyst.json.JsonInferSchema.$anonfun$infer$2(JsonInferSchema.scala:66)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:484)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:490)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.TraversableOnce.reduceLeft(TraversableOnce.scala:190)
	at scala.collection.TraversableOnce.reduceLeft$(TraversableOnce.scala:183)
	at scala.collection.AbstractIterator.reduceLeft(Iterator.scala:1429)
	at scala.collection.TraversableOnce.reduceLeftOption(TraversableOnce.scala:208)
	at scala.collection.TraversableOnce.reduceLeftOption$(TraversableOnce.scala:207)
	at scala.collection.AbstractIterator.reduceLeftOption(Iterator.scala:1429)
	at scala.collection.TraversableOnce.reduceOption(TraversableOnce.scala:215)
	at scala.collection.TraversableOnce.reduceOption$(TraversableOnce.scala:215)
	at scala.collection.AbstractIterator.reduceOption(Iterator.scala:1429)
	at org.apache.spark.sql.catalyst.json.JsonInferSchema.$anonfun$infer$1(JsonInferSchema.scala:81)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:863)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:863)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

### Why are the changes needed?

bugfix

### Does this PR introduce _any_ user-facing change?

yes, ignoreCorruptFiles will ignore JSON files corrupted

### How was this patch tested?

new test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala [hunk 1]
```diff
@@ -53,7 +55,8 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
    */
   def infer[T](
       json: RDD[T],
-      createParser: (JsonFactory, T) => JsonParser): StructType = {
+      createParser: (JsonFactory, T) => JsonParser,
+      ignoreCorruptFiles: Boolean): StructType = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:59]: Can we get it from `SQLConf.get`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 2]
```diff
@@ -2435,6 +2435,17 @@ abstract class JsonSuite
     }
   }
 
+  test(s"infer json schema respect") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:2438]: Let's add a JIRA number as a prefix (and remove `s`).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala [hunk 3]
```diff
@@ -68,13 +71,19 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
             Some(inferField(parser))
           }
         } catch {
-          case  e @ (_: RuntimeException | _: JsonProcessingException) => parseMode match {
-            case PermissiveMode =>
-              Some(StructType(Seq(StructField(columnNameOfCorruptRecord, StringType))))
-            case DropMalformedMode =>
+          case e @ (_: RuntimeException | _: JsonProcessingException | _: IOException) =>
+            if (ignoreCorruptFiles) {
+              logWarning(s"Skipped the corrupted file: $row", e)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:76]: Can we keep the log consistent with other places? e.g.) Skipped the rest of the content in ...

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:77]: here we use it for inferring schema, not all the content.

At avro side `org.apache.spark.sql.avro.AvroUtils#inferAvroSchemaFromFiles`, we say
```
Skipped the footer in the corrupted file: ...
```

How about we print `Skipped inferring schema from the corrupted file: ...`

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:77]: Also should we maybe exclude `RuntimeException` for now? I think we intentionally throw `RuntimeException` for some places like:

```
    case token =>
      // We cannot parse this token based on the given data type. So, we throw a
      // RuntimeException and this exception will be caught by `parse` method.
      throw QueryExecutionErrors.failToParseValueForDataTypeError(parser, token, dataType)
```

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:77]: That's fine. but just to make extra sure `row` is being printed, not the file name. Maybe we should handle it in a higher level?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala]: seems very hard to do it at a higher level

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala [hunk 4]
```diff
@@ -68,21 +73,25 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
...
           }
         } catch {
-          case  e @ (_: RuntimeException | _: JsonProcessingException) => parseMode match {
-            case PermissiveMode =>
-              Some(StructType(Seq(StructField(columnNameOfCorruptRecord, StringType))))
-            case DropMalformedMode =>
-              None
-            case FailFastMode =>
-              throw QueryExecutionErrors.malformedRecordsDetectedInSchemaInferenceError(e)
+          case e @ (_: JsonProcessingException | _: IOException) if ignoreCorruptFiles =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:76]: Should we maybe just catch `CharConversionException` to match with `JacksonParser`?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala]: sgtm, updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala [hunk 5]
```diff
@@ -68,21 +73,27 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
...
-              None
-            case FailFastMode =>
-              throw QueryExecutionErrors.malformedRecordsDetectedInSchemaInferenceError(e)
+          case e @ (_: JsonProcessingException | _: CharConversionException)
+              if ignoreCorruptFiles =>
+            logWarning("Skipped inferring json schema from the corrupted data: " +
+              s"${row.asInstanceOf[InternalRow].getString(0)}", e)
+            None
+          case e @ (_: RuntimeException | _: JsonProcessingException |
+                    _: CharConversionException) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:82]: @yaooqinn can we just don't respect `ignoreCorruptFiles` here for now, and match here with `JacksonParser`:

https://github.com/apache/spark/blob/186815be1c0832e2f976faf6f2d5c31e42eebf1a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala#L495

https://github.com/apache/spark/blob/186815be1c0832e2f976faf6f2d5c31e42eebf1a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala#L490

meaning that here we will catch two more exceptions of `CharConversionException` (if encoding is not specified) and `MalformedInputException`.

I think JSON datasource itself doesn't really support `ignoreCorruptFiles` (not only JSON schema inference). For example, if other exceptions happen or `encoding` is specified, I think the exception will be thrown.

**Comment by HyukjinKwon** [comment]
cc @MaxGekk would you mind taking a look please when you find some time?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala (hunk 5 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala]: thanks, can you take a look again to see if I did get your point

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala [hunk 6]
```diff
@@ -68,13 +71,15 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
...
         } catch {
-          case  e @ (_: RuntimeException | _: JsonProcessingException) => parseMode match {
-            case PermissiveMode =>
-              Some(StructType(Seq(StructField(columnNameOfCorruptRecord, StringType))))
-            case DropMalformedMode =>
-              None
-            case FailFastMode =>
-              throw QueryExecutionErrors.malformedRecordsDetectedInSchemaInferenceError(e)
+          case e @ (_: RuntimeException | _: JsonProcessingException |
+                    _: CharConversionException | _: MalformedInputException) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:75]: Let's separate CharConverrsionException like JacksonParser, and change its cause like:
https://github.com/apache/spark/blob/186815be1c0832e2f976faf6f2d5c31e42eebf1a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala#L495-L501

BTW, the indentation below here seems wrong

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 7]
```diff
@@ -2435,6 +2435,24 @@ abstract class JsonSuite
     }
   }
 
+  test(s"SPARK-36477: infer json schema respect spark.sql.files.ignoreCorruptFiles") {
+    withSQLConf((SQLConf.IGNORE_CORRUPT_FILES.key, "true")) {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:2439]: I think we don't need this anymore. Let's also change testtitle, pr description etc.

**Comment by HyukjinKwon** [comment]
Looks good otherwise.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:75]: cases like https://github.com/apache/spark/pull/33706/files#diff-587e4c03ef08aa4964d4409ae33954735a9c090727dd7b5a918892cd3ed0814dR2450 seems not to have a chance to add JSON options

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala (hunk 7 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:2439]: yes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:75]: And when reading from files, `CharConversionException` can be not related to the value, but also something else like compression. Maybe we shall handle it differently from the JacksonParser above

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala [hunk 8]
```diff
@@ -68,7 +70,8 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
             Some(inferField(parser))
           }
         } catch {
-          case  e @ (_: RuntimeException | _: JsonProcessingException) => parseMode match {
+          case e @ (_: RuntimeException | _: JsonProcessingException |
+                    _: CharConversionException | _: MalformedInputException) => parseMode match {
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:74]: copied from org.apache.spark.sql.catalyst.json.JacksonParser#parse

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:74]: Yeah, it's to match with JacksonParser one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala (hunk 6 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala]: It actually does via rawParser, e.g.) https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala#L105-L111

Let's match it first and fix it in a different place.

**Comment by HyukjinKwon** [comment]
sorry @yaooqinn for a late response - it slipped through my fingers. I left one comment: https://github.com/apache/spark/pull/33706#discussion_r693421205

**Pull request closed without merging**